src/utils.ts

```diff
@@ -1,41 +1,141 @@
 import React from "react";
-import { SubmissionMetadata, TEMPLATES } from "./api";
+import { SampleData, SubmissionMetadata, TemplateName, TEMPLATES } from "./api";
 import { SlotDefinition } from "./linkml-metamodel";
 
+/**
+ * Get the templates associated with a submission.
+ *
+ * This function is mainly useful to ease the transition between the old format where the
+ * `packageName` field was a string and the new format where it is an array of strings.
+ * This function always returns an array of strings.
+ *
+ * @param submission
+ */
+export function getSubmissionTemplates(
+  submission?: SubmissionMetadata,
+): TemplateName[] {
+  if (!submission) {
+    return [];
+  }
+  const { packageName } = submission.metadata_submission;
+  if (typeof packageName === "string") {
+    if (packageName === "") {
+      return [];
+    } else {
+      return [packageName];
+    }
+  }
+  return packageName;
+}
+
 export interface GetSubmissionSamplesOptions {
   createSampleDataFieldIfMissing?: boolean;
 }
+
+/**
+ * Get the samples associated with a submission.
+ *
+ * Returns an object where the keys are the templates and the values are arrays of sample data
+ * objects for that template. If the `createSampleDataFieldIfMissing` option is set to true, it
+ * will modify the submission object to add missing template fields, defaulting to an empty array.
+ *
+ * @param submission
+ * @param options
+ */
 export function getSubmissionSamples(
   submission?: SubmissionMetadata,
   options: GetSubmissionSamplesOptions = {},
-) {
+): Partial<Record<TemplateName, SampleData[]>> {
   if (!submission) {
-    return [];
+    return {};
+  }
+  const samples: Partial<Record<TemplateName, SampleData[]>> = {};
+  const templates = getSubmissionTemplates(submission);
+  templates.forEach((template) => {
+    const sampleDataSlot = TEMPLATES[template]?.sampleDataSlot;
+    if (!sampleDataSlot) {
+      return;
+    }
+    if (
+      !(sampleDataSlot in submission.metadata_submission.sampleData) &&
+      options.createSampleDataFieldIfMissing
+    ) {
+      submission.metadata_submission.sampleData[sampleDataSlot] = [];
+    }
+    samples[template] =
+      submission.metadata_submission.sampleData[sampleDataSlot] || [];
+  });
+  return samples;
+}
+
+/**
+ * Get the total number of samples associated with a submission across all templates.
+ *
+ * @param submission
+ */
+export function getSubmissionSamplesCount(
+  submission?: SubmissionMetadata,
+): number {
+  if (submission === undefined) {
+    return 0;
   }
-  const environmentalPackageName = submission.metadata_submission.packageName;
-  const sampleDataField = environmentalPackageName
-    ? TEMPLATES[environmentalPackageName].sampleDataSlot
-    : undefined;
-  if (!sampleDataField) {
+  let count = 0;
+  const templates = getSubmissionTemplates(submission);
+  templates.forEach((template) => {
+    const sampleDataSlot = TEMPLATES[template]?.sampleDataSlot;
```

Comment:
This is a follow-up comment to the first one I left. This is the final file in the diff. After having reviewed the other files in the diff, I'd prefer `template` be named `templateName`. Given what `template` contains (i.e. the name of a template), I find `TEMPLATES[templateName]` easier to mentally process than `TEMPLATES[template]`.